### ISSUE TITLE
Extremely slow performance with StableHLO

### ISSUE THREAD

**Issue by SamKG** [pytorch/xla#5658]
## 🐛 Bug

StableHLO performance currently seems to be 2 orders of magnitude worse than the normal XLA flow.

## To Reproduce
Please try the following script:

```python
import os
import timeit
import torch
import torchvision
import torch_xla.core.xla_model as xm
from torch.export import export
from torch_xla.stablehlo import exported_program_to_stablehlo

os.environ["PJRT_DEVICE"] = "GPU"
os.environ["GPU_NUM_DEVICES"] = "1"

def measure_perf(fn, *args, n_iters=10, **kwargs):
    times = []
    for i in range(n_iters):
        start_t = timeit.default_timer()
        result = fn(*args, **kwargs)
        end_t = timeit.default_timer()
        times.append(end_t - start_t)
    return times

# Benchmark only one model to isolate the issue
model_class = torchvision.models.resnet18
model = model_class()
input_data = torch.randn(128, 3, 224, 224)

results = {}

with torch.inference_mode():

    xla_cuda = xm.xla_device(devkind="GPU")
    results["CUDA (XLA)"] = measure_perf(model.to(xla_cuda), input_data.to(xla_cuda))
    exported = export(model.to(xla_cuda), (input_data.to(xla_cuda), ))
    model_stablehlo = exported_program_to_stablehlo(exported)
    results["CUDA (XLA, StableHLO)"] = measure_perf(model_stablehlo, input_data.to(xla_cuda))

print(results)

```

1. Run above script 
2. Observe runtime gap

The runtimes of the above are as follows (on A100 40GB):

```
   CUDA (XLA)  CUDA (XLA, StableHLO)     Model
0    0.004488               3.045514  ResNet18 <------ first run (maybe some compilation or init overhead?)
1    0.006988               0.934498  ResNet18
2    0.006405               0.931537  ResNet18
3    0.006394               0.961547  ResNet18
4    0.006419               0.945377  ResNet18
5    0.006529               0.934150  ResNet18
6    0.006540               0.934026  ResNet18
7    0.006716               0.939513  ResNet18
8    0.006838               0.933823  ResNet18
9    0.006815               0.931438  ResNet18
```

## Expected behavior

Similar or equal performance between StableHLO and the default execution path

## Environment

 - Reproducible on XLA backend [CPU/TPU]:
 - torch_xla version: nightly

**Comment by JackCaoG** [comment]
@qihqi can you take a look?

**Comment by qihqi** [comment]
Hi @SamKG 

So when you evaluate a model using XLA tensor (i.e. Lazy tensor), you are only creating the lazy tensor graph without actually performing any computation. Computation will be triggered when the value of the tensor is used (i.e. `print(tensor)`) or forced to materialize (i.e. via `xm.mark_step()`)

Adding `print(result)` next to `result = fn(*args, **kwargs)`; I got them to be the similar (3.9 vs 4s on TPU,).
StableHLO version is still lower because it needs to convert stableHLO back into HLO to run, but they are in the same order of magnitude.

**Comment by SamKG** [comment]
> Hi @SamKG
> 
> So when you evaluate a model using XLA tensor (i.e. Lazy tensor), you are only creating the lazy tensor graph without actually performing any computation. Computation will be triggered when the value of the tensor is used (i.e. `print(tensor)`) or forced to materialize (i.e. via `xm.mark_step()`)
> 
> Adding `print(result)` next to `result = fn(*args, **kwargs)`; I got them to be the similar (3.9 vs 4s on TPU,). StableHLO version is still lower because it needs to convert stableHLO back into HLO to run, but they are in the same order of magnitude.

Thanks for the reply!

I wasn't able to replicate the performance gains; here is my modified benchmark function:
```python
def measure_perf(fn, *args, n_warmup=5, n_iters=10, **kwargs):
    for i in range(n_warmup):
        fn(*args, **kwargs)
    times = []
    for i in range(n_iters):
        result = fn(*args, **kwargs) # <---------------   moved lazy tensor graph here
        start_t = timeit.default_timer()
        xm.mark_step(wait=True) #    <----------------   added this
        end_t = timeit.default_timer()
        times.append(end_t - start_t)
    return times
```

can you clarify if my benchmark function is correct?

**Comment by tzstoyanov** [comment]
Hi @SamKG,
Do you have any progress analyzing that performance problem? I'm considering using StableHlo inference in my project, but my initial tests show the same performance issue, it is even worse in my measurements. I tested Gpt2 running on Nvidia Quadro T1000 @ 4G vram, run a loop of 20 iterations. Noticed that the inference of the original model is extremely slow on the first few iterations, and after that runs much faster; while the inference of the StableHlo model does not depend on the iteration count - it is constantly slow.

**Comment by JackCaoG** [comment]
the slowdown in first few steps most likely due to compilation, you can enable persistent cache following https://github.com/pytorch/xla/blob/master/API_GUIDE.md#compilation-caching .

**Comment by tzstoyanov** [comment]
Thank @JackCaoG, I didn't know about that config. I enabled the persistent cache, it has big improvement on original model inference - the first iterations take half the time. There is also a speed up of the StableHlo inference, but this is not so significant: **0.28548  tokens/s**,  still far from the original model.
Is this a known issue?

**Comment by JackCaoG** [comment]
When you say `StableHLO inference` what do you mean? Do you mean turning on `XLA_STABLEHLO_COMPILE` flags? or do you export the model to `StableHLO` format and serve it directly(how?) I assume when you say original you just runs the pytorch/xla model and add `mark_step` at the step end. Are you using `torch.compile` at any setup?

**Comment by tzstoyanov** [comment]
@JackCaoG  thanks for looking into that, my test program is similar to the one from this PR, with small modifications to run LLM. I use `torch.export.export` + `exported_program_to_stablehlo`:

```python
import os
import time
from tqdm import tqdm
import torch
import torch_xla.core.xla_model as xm
from transformers import AutoModelForCausalLM, AutoTokenizer
from torch_xla.stablehlo import exported_program_to_stablehlo, StableHLOExportOptions

def run_test(model, intensor, count=5):
  times = []
  for i in tqdm(range(count)):
    tstamp = time.time()
    res = model(intensor)
    xm.mark_step(wait=True)
    times.append(time.time() - tstamp)
  return times

def run_stablehlo_inference(model_path, prompt):
  device = xm.xla_device(devkind="GPU")
  tokenizer = AutoTokenizer.from_pretrained(model_path)
  model = AutoModelForCausalLM.from_pretrained(model_path)
  model = model.to(device)
  ptokens = tokenizer(prompt, return_tensors='pt')
  intensor = ptokens['input_ids']
  intensor = intensor.to(device)
  intensor.requires_grad = False
  exported = torch.export.export(model, (intensor,))
  options = StableHLOExportOptions()
  options.override_tracing_arguments = (intensor,)
  shlo = exported_program_to_stablehlo(exported, options)
  with torch.no_grad():
    res_xla = run_test(model, intensor) 
    print(res_xla)
    res_shlo = run_test(shlo, intensor)
    print(res_shlo)  

if __name__ == '__main__':
  os.environ["PJRT_DEVICE"] = "CUDA"
  os.environ["GPU_NUM_DEVICES"] = "1"
  run_stablehlo_inference("gpt2", "Who are you?")
```

**Comment by JackCaoG** [comment]
Hmm. I took a quick look at `exported_program_to_stablehlo` it returns a callable and can be executed on XLA device. @qihqi I think what happened is that when executing that shlo it got converted back to HLO since our runtime expects HLO, can you confirm?

**Comment by qihqi** [comment]
Hi, @tzstoyanov

Your observation on running StableHLO using the stock experience is slow independent of number of iterations sounds good to me. The reason is that, we execute the stablehlo bytecode at this line: https://github.com/pytorch/xla/blob/master/torch_xla/stablehlo.py#L80

This function actually does 2 things: it compiles the bytecode, then it executes that bytecode. It does not save compiled artifact for later use. Therefore, the first run and later run is at the same speed.

Originally we added functionality to run stablehlo for testing. As before that we used to execute stablehlo using jax or tensorflow because this is intended to be an export experience; where after export the users take the stablehlo bytecode to somewhere else for actual execution (or further lowering into other things).

To make this as a suitable for your use case, it need to be modified to cache the compiled artifact. 

Another alternative would be to export it into a `tf.saved_model` and use tensorflow for inference.

**Comment by JackCaoG** [comment]
Thanks @qihqi for the context. I took another look at the code, I think compilation is actually cached by the `persistent cache`, however to execute the stableHLO we first converted it back to HLO in https://github.com/pytorch/xla/blob/dc17f9c08851761916e00ec700bb6aa20d6dc2f9/torch_xla/csrc/xla_graph_executor.cpp#L871-L877

which adds some overhead.

@tzstoyanov I guess my questionis what's your use case? XLA:TPU and XLA:GPU both takes HLO as input, whether you execute it directly or convert it first to stableHLO it go through the exact same compiler and optimizations. The stableHLO export is more useful when you have your custom compiler pass you want to run.


Going back to the inference, I think you will get better throughput by using `torch.compile` since it will even bypass the tracing. Some example in https://github.com/pytorch/xla/blob/master/docs/dynamo.md#inference.
